dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+Another intesting thing is record and its type. For record, we unify given signatures of two records, and then their field values. For types of records we only unify types.
+```haskell
+unify (Record ds) (Record gs) = do
+  case zipDecls ds gs of
+    Just quads -> for_ quads \(tn, tm, n, m) -> do
+      unify tn tm  -- unify inferrent types of fields
+      unify  n  m  -- unify the fields themselves
+
+    Nothing -> do
+      die $ Mismatch (Record ds) (Record gs)
+
+unify (Product tds) (Product tgs) = do
+  case zipTDecls tds tgs of
+    Just pairs -> for_ pairs \(n, m) -> do
+      unify n m
+
+    Nothing -> do
+      die $ Mismatch (Product tds) (Product tgs)
+```
+
+Sadly, we can do nothing with _projections_. It is possible to have a type expression `a.b`, where `a` is a bound variable and `b` is a field name. It is not reasonable to unify `a ~ b` for `a.x ~ b.x`, because it is possible to have latter without the former. We will consider this an error. Technically, there are ways around it - by handling special cases - but that's not a general solution.
+```haskell
+unify a@Match {} b = die $ ProjectionL a b
+unify a b@Match {} = die $ ProjectionR a b
+
+unify a@Get {} b = die $ ProjectionL a b
+unify a a@Get {} = die $ ProjectionR a b
+```
+
+The `let`-expression of both kinds should not enter unifier at all. They can always be normalised to non-`let`-expressions.
+```haskell
+unify Let {} _ = error "unifier: left argument is a let-expression"
+unify _ Let {} = error "unifier: right argument is a let-expression"
+
+unify LetRec {} _ = error "unifier: left argument is a let-expression"
+unify _ LetRec {} = error "unifier: right argument is a let-expression"
+```
+
+And the last case is fired after all other fail, which means that types don't match.
+```haskell
+unify a b = die $ Mismatch a b
+```
+
+## Normalization
+
+Before we can start with inference, there is another matter to discuss. In simpler type systems the type are always in their normal form, there is no abstraction that disobeys `f x ~ g x -> f ~ g` rule, which means that there is no type lambdas.
+
+We have no such restrictions here. There _can_ be lambdas, which gives us a need to evaluate type expression before using them in type checks.
+
+However, the normalization is easy, since the expression to be normalised will be typechecked _before_ it enters normaliser, which kinda gives us moral right to ignore all possible "runtime" failures in normalisation.
+
+I will mostly skip the description of it. The single problem I met is a representation of types. I made a separate constructor for that:
+```haskell
+Axiom :: Name -> self -> AST_ self
+```
+This constructor can only be inserted by normalizer, by evaluating, for instance
+```haskell
+let
+  data Bool where
+    | True  : Bool
+    | False : Bool
+in
+  <ctx>
+```
+
+The names `Bool`, `True` and `False` will be replaced in `ctx` with (in pseudosyntax)
+
+```
+axiom "True" Bool
+axiom "True" Bool
+axiom "Bool" Type
+```
```

Comment:
Are these something like inference rules for values?

Reply:
Yes, they are runtime constructor reps, with the type.